### PR TITLE
nitrokey-app2: switch to pypaBuildHook and fix the build

### PR DESCRIPTION
**Pull request by panicgh**
## Description of changes

nitrokey-app2 currently fails to build.
Successfully tested with a nk3 and pynitrokey 0.4.39.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/security/nitrokey-app2/default.nix [hunk 1]
```diff
@@ -17,11 +19,19 @@ python3.pkgs.buildPythonApplication rec {
     hash = "sha256-VyhIFNXxH/FohgjhBeZXoQYppP7PEz+ei0qzsWz1xhk=";
   };
 
+  # https://github.com/Nitrokey/nitrokey-app2/issues/152
+  postPatch = ''
+    substituteInPlace pyproject.toml --replace "pynitrokey ==" "pynitrokey >="
+  '';
```
Comment (999eagle) [inline pkgs/tools/security/nitrokey-app2/default.nix:25]:
```suggestion
  pythonRelaxDeps = [
    "pynitrokey"
  ];
```

File: pkgs/tools/security/nitrokey-app2/default.nix [hunk 2]
```diff
@@ -17,11 +19,19 @@ python3.pkgs.buildPythonApplication rec {
...
+
+  # The pyproject.toml file seems to be incomplete and does not generate
+  # resources (i.e. run pyrcc5 and pyuic5) but the Makefile does.
   preBuild = ''
     make build-ui
   '';
 
   nativeBuildInputs = with python3.pkgs; [
+    flit-core
```
Comment (999eagle) [inline pkgs/tools/security/nitrokey-app2/default.nix:34]:
```suggestion
    flit-core
    pythonRelaxDepsHook
```

**Review by 999eagle — CHANGES_REQUESTED** [review]
I'd recommend using the `pythonRelaxDepsHook` to make the dependency relaxation more obvious and robust.

Apart from that LGTM and works with my NK3.

File: pkgs/tools/security/nitrokey-app2/default.nix (hunk 1 above)
Comment (panicgh) [reply pkgs/tools/security/nitrokey-app2/default.nix:25]: I tried this, but it didn't work.

Comment (999eagle) [reply pkgs/tools/security/nitrokey-app2/default.nix:25]: Ah I see, I tried it myself now. The [`pythonRelaxDepsHook`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/hooks/python-relax-deps-hook.sh) is executed in the `postBuild` phase and only modifies the dependencies in the built distribution. I'm not sure if there is a better way than this substitution but it works fine.

**Review by 999eagle — APPROVED** [review]

**Comment by panicgh** [comment]
I added a comment explaining the issue with `pythonRelaxDepsHook` and added `meta.changelog`.
@999eagle please review again.

**Review by 999eagle — APPROVED** [review]

**Review by fabaff — APPROVED** [review]
Looks good to me.

**Pull request merged**